Tournament_Calculator.py

```diff
@@ -0,0 +1,323 @@
+
+
+
+
+#Swiss Pairings Logic
+def generate_pairings(players, previous_pairings=None, results=None):
+    pairings = []
+    if previous_pairings is None:
+        previous_pairings = set()  # Initialize an empty set for previous pairings
+    if results is None:
+        results = {}  # Initialize an empty dictionary for results
+    
+    # Group players based on their current results
+    result_groups = {}
+    for player in players:
+        result = results.get(player, 0)
+        if result not in result_groups:
+            result_groups[result] = []
+        result_groups[result].append(player)
+
+
+    
+    # Sort result groups based on results
+    sorted_result_groups = sorted(result_groups.items(), key=lambda x: x[0], reverse=True)
+
+
+    # Pair players with similar results
+    for i, (result, group) in enumerate(sorted_result_groups):
+        if len(group) % 2 != 0:
+            # If there's an odd number of players in the group, try to pair with the highest-ranked player in the next lowest group
+            next_group_index = i + 1
+            while next_group_index < len(sorted_result_groups):
+                next_result, next_group = sorted_result_groups[next_group_index]
+                if len(next_group) > 0:
+                    odd_player = group[0]
+                    random.shuffle(next_group)
+                    pair = (group[0], next_group[0])
+                    while pair in previous_pairings or pair[::-1] in previous_pairings:
+                        random.shuffle(next_group)
+                        random.shuffle(group)# Shuffle players again to try different pairings
+                        pair = (group[0], next_group[0])
+                        if len(next_group) == 3:
+                            pair2 = (next_group[1], next_group[2])
+                            while pair2 in previous_pairings or pair[::-1] in previous_pairings:
+                                random.shuffle(next_group)
+                                pair = (group[0], next_group[0])
+                            if len(group) == 3:
+                                pair3 = ((group[1], group[2]))
+                                while pair3 in previous_pairings or pair[::-1] in previous_pairings:
+                                    random.shuffle(group)
+                                    pair = (group[0], next_group[0])
+                    odd_player = group.pop()
+                    paired_player = next_group.pop(0)
+                    pairings.append((odd_player, paired_player))

```

Comment:
An interesting design challenge you could give yourself would be to see if you could simplify this code by writing some special classes that hide some of these operations. Some example ideas (I haven't thought these through in detail so don't take this as gospel):

- A `PlayerPairing` class that holds the two players, overloads the equality operator so they can be compared (https://www.pythonmorsels.com/overloading-equality-in-python/), and has a function to decide the winner between the two
- A `GameRound` class which builds + holds all of the pairings for that round and can pass the results forward to the creation of the next round. It could have a `HasPairing` method to check if a given player pairing happened already.

Lmk if you're not super familiar with classes or object-oriented design and I'll throw some articles at you. In that case I would doubly emphasize that you should try this, it might be overkill for this specific example but it would be great learning.